kiwi/system/prepare.py

```diff
@@ -355,6 +366,27 @@ def delete_packages(self, manager, packages, force=False):
                     'Package deletion failed: %s' % format(e)
                 )
 
+    def delete_packages_without_configured_repos(self, packages, force=False):
+        """
+        This is a wrapper of delete_packages that does not require a package
+        manager instance. It performs exactly the same code using a dummy
+        package manager instance created on the fly.
+
+        Repositories configuration is unneeded for packages removal, thus
```

Comment:
Don't understand this. The force deletion doesn't need a repo setup but the uninstall imho does. In the zypper case we call it with the '-u' flag which does a backward resolver task looking which packages are no longer needed when the package to become deleted gets removed. imho that requires access to the configured repo metadata. I can't tell how the other package manager does it but I would say a clean dep removal task needs the repo metadata

In the context where this method is called you also have access to an instance of PackageManager, I wonder why this method is needed. Unless I completely overlook something this feels weird

Reply:
Agree that feels weird, I did not find a better way to do it. I am totally opened to improve or modify it.

> In the zypper case we call it with the '-u' flag which does a backward resolver task looking which packages are no longer needed when the package to become deleted gets removed. imho that requires access to the configured repo metadata.

Well, I am not sure about that, why would it need it? I'd say rpmdb should be enough IMHO. In my tests it worked fine without any repo configuration.

Anyway, the motivation of it comes from https://github.com/SUSE/kiwi/blob/348c0193245fd637c3ab841ef1ec34bd6ae5b56c/kiwi/tasks/system_prepare.py#L218 note that the manager instance (with build time repositories configured) is deleted before running `config.sh`. Thus after `config.sh` we don't have anymore the package_manager instance we used for the install phase. We could probably recreate it (by reconfiguring buildtime repos again), I just haven't considered it since it also looked to me weird and not as safe as do not touch any repo configuration.

Thoughts?

Reply:
> Well, I am not sure about that, why would it need it? I'd say rpmdb should be enough

you are right, looking closer to how zypper deletes packages this works for the rpmdb case. If we can guarantee that this applies to all other package managers too I'm ok. It's most likely that this is the case as they all maintain a database and the package manager sits on top.

Reply:
Well, the tests I did were all successful with dnf, yum and zypper. With APT the behavior was different, but as @Conan-Kudo noted [here](https://github.com/SUSE/kiwi/pull/680#issuecomment-378209931), this is a difference on the package manager design, APT can't constraint the removal of unneeded dependencies to a specific package, it performs a global clean up. At the time I tested #681 I did all the tests together with #680, where also no repositories were configured and I had no issues with any of the package managers (apart of the APT that behaves different).

Reply:
ok :+1: